ci/tcli/tcli.go

```diff
@@ -0,0 +1,328 @@
+)
+
+var (
+	_ runnable = &ContainerRunner{}
+	_ runnable = &HostRunner{}
+)
+
+type runnable interface {
+	Run(ctx context.Context, command string) *Assertable
+	RunCmd(cmd *exec.Cmd) *Assertable
+	io.Closer
+}
+
+// ContainerConfig describes the ContainerRunner configuration schema for initializing a testing environment
+type ContainerConfig struct {
+	Name       string
+	Image      string
+	BindMounts map[string]string
+}
+
+func mountArgs(m map[string]string) (args []string) {
+	for src, dest := range m {
+		args = append(args, "--mount", fmt.Sprintf("type=bind,source=%s,target=%s", src, dest))
+	}
+	return args
+}
+
+func preflightChecks() error {
+	_, err := exec.LookPath("docker")
+	if err != nil {
+		return xerrors.Errorf(`"docker" not found in $PATH`)
+	}
+	return nil
+}
+
+// ContainerRunner specifies a runtime container for performing command tests
+type ContainerRunner struct {
+	name string
+	ctx  context.Context
+}
+
+// NewContainerRunner starts a new docker container for executing command tests
+func NewContainerRunner(ctx context.Context, config *ContainerConfig) (*ContainerRunner, error) {
+	if err := preflightChecks(); err != nil {
+		return nil, err
+	}
+
+	args := []string{
+		"run",
+		"--name", config.Name,
+		"-it", "-d",
+	}
+	args = append(args, mountArgs(config.BindMounts)...)
+	args = append(args, config.Image)
+
+	cmd := exec.CommandContext(ctx, "docker", args...)
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return nil, xerrors.Errorf(
+			"failed to start testing container %q, (%s): %w",
+			config.Name, string(out), err)
+	}
+
+	return &ContainerRunner{
+		name: config.Name,
+		ctx:  ctx,
+	}, nil
+}
+
+// Close kills and removes the command execution testing container
+func (r *ContainerRunner) Close() error {
+	cmd := exec.CommandContext(r.ctx,
+		"sh", "-c", strings.Join([]string{
+			"docker", "kill", r.name, "&&",
+			"docker", "rm", r.name,
+		}, " "))
+
+	out, err := cmd.CombinedOutput()
+	if err != nil {
+		return xerrors.Errorf(
+			"failed to stop testing container %q, (%s): %w",
+			r.name, string(out), err)
+	}
+	return nil
+}
+
+// Run executes the given command in the runtime container with reasonable defaults.
+// "command" is executed in a shell as an argument to "sh -c".
+func (r *ContainerRunner) Run(ctx context.Context, command string) *Assertable {
+	cmd := exec.CommandContext(ctx,
+		"docker", "exec", "-i", r.name,
+		"sh", "-c", command,
+	)
+
+	return &Assertable{
+		cmd:   cmd,
+		tname: command,
+	}
+}
+
+// RunCmd lifts the given *exec.Cmd into the runtime container
+func (r *ContainerRunner) RunCmd(cmd *exec.Cmd) *Assertable {
+	path, _ := exec.LookPath("docker")
+	cmd.Path = path
+	command := strings.Join(cmd.Args, " ")
+	cmd.Args = append([]string{"docker", "exec", "-i", r.name}, cmd.Args...)
+
+	return &Assertable{
+		cmd:   cmd,
+		tname: command,
+	}
+}
+
+// HostRunner executes command tests on the host, outside of a container
+type HostRunner struct{}
```

Comment:
Do you foresee us using `HostRunner` in the coder-cli tests?

Reply:
Not at the moment, but I do think it has a place in `tcli` as an escape hatch.